File: AdventOfCode2021/FSharp/Day1.fs
module FSharp.Day1

open System.IO
open FSharp.Core

let parseFileInput (path: string) : string = File.ReadAllText(path)

let parseInput (text: string) : int array =
    text.Split("\r\n")
    |> Array.map (fun it -> it |> int)

type State =
    { last: int option
      total: int }
    static member empty = { last = None; total = 0 }


let part1 (text: string) : int =
    let input = parseInput text

    let folder (state: State) (current: int) : State =
        match state.last with
        | None -> { state with last = Some current }
        | Some last ->
            match current > last with
            | true -> { last = Some current; total = state.total + 1 }
            | false -> { state with last = Some current }

    input
    |> Array.fold folder State.empty
    |> fun it -> it.total


let part2 (text: string) : int =
    let input = parseInput text

    let folder (state: State) (current: int []) : State =
        let sum = current |> Array.sum
        
        match state.last with
        | None -> { state with last = Some sum }
        | Some last ->
            match sum > last with
            | true -> { last = Some sum; total = state.total + 1 }
            | false -> { state with last = Some sum }

    input
    |> Array.windowed 3
    |> Array.fold folder State.empty
    |> fun it -> it.total


module Test =
    open Xunit

    let smallInput =
        @"199
200
208
210
200
207
240
269
260
263"

    [<Fact>]
    let ``Part 1 Test`` () =
        let expected = 7
        let actual = part1 smallInput
        Assert.Equal(expected, actual)

    [<Fact>]
    let ``Part 2 Test`` () =
        let expected = 5
        let actual = part2 smallInput
        Assert.Equal(expected, actual)

File: AdventOfCode2021/CSharp/Day1.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace AdventOfCode2021
{
    public class Day1
    {
        public int Part1(string input)
        {
            var tokens = input.Split("\r\n").Select(int.Parse).ToList();
            var increaseCount = 0;
            var last = tokens.First();
            foreach (var token in tokens.Skip(1))
            {
                if (token > last)
                    increaseCount++;
                last = token;
            }

            return increaseCount;
        }

        public int Part2(string input)
        {
            var tokens = input.Split("\r\n").Select(int.Parse);
            var windowed = tokens.Window(3).ToList();
            var increaseCount = 0;
            var last = windowed.First().Sum();
            foreach (var window in windowed.Skip(1))
            {
                var sum = window.Sum();
                if (sum > last)
                    increaseCount++;
                last = sum;
            }

            return increaseCount;
        }
    }
    
    [TestClass]
    public class Day1Tests
    {
        private string _smallInput = @"199
200
208
210
200
207
240
269
260
263";
        
        [TestMethod]
        public void TestPart1()
        {
            var expected = 7;
            var actual = new Day1().Part1(_smallInput);
            Assert.AreEqual(expected, actual);
        }
        
        [TestMethod]
        public void TestPart2()
        {
            var expected = 5;
            var actual = new Day1().Part2(_smallInput);
            Assert.AreEqual(expected, actual);
        }
    }
}